File: customer-java-webflux/src/main/java/com/prez/api/CustomJwtTokenHandler.java
package com.prez.api;

import static org.springframework.http.HttpStatus.UNAUTHORIZED;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.web.reactive.function.server.HandlerFilterFunction;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Mono;

public class CustomJwtTokenHandler implements HandlerFilterFunction<ServerResponse, ServerResponse> {

  private static final Logger LOGGER = LoggerFactory.getLogger(CustomJwtTokenHandler.class);

  @Override
  public Mono<ServerResponse> filter(ServerRequest request, HandlerFunction<ServerResponse> next) {
    return request.principal()
        .doOnSuccess(principal
            -> LOGGER.debug("Principal type : {} - value : {}", principal.getClass().getName(), principal))
        .ofType(JwtAuthenticationToken.class)
        .switchIfEmpty(Mono.error(new ResponseStatusException(UNAUTHORIZED, "No principal")))
        // Log just to see if we get the value of the custom azp attribute, but imagine we are doing something meaningful here
        .doOnSuccess(principal ->
            LOGGER.info("Principal info : name={}, azp={}",
                principal.getName(),
                principal.getTokenAttributes().get("azp")))
        .then(next.handle(request));
  }
}

File: customer-kotlin-webflux/src/main/kotlin/com/prez/api/CustomJwtTokenHandler.kt
package com.prez.api

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.web.reactive.function.server.HandlerFilterFunction
import org.springframework.web.reactive.function.server.HandlerFunction
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Mono

class CustomJwtTokenHandler : HandlerFilterFunction<ServerResponse, ServerResponse> {

  private val logger = LoggerFactory.getLogger(CustomJwtTokenHandler::class.java)

  override fun filter(request: ServerRequest, next: HandlerFunction<ServerResponse>): Mono<ServerResponse> {
    return request.principal()
      .doOnSuccess {
        logger.debug("Principal type : {} - value : {}", it.javaClass.name, it)
      }
      .ofType(JwtAuthenticationToken::class.java)
      .switchIfEmpty(Mono.error(ResponseStatusException(HttpStatus.UNAUTHORIZED, "No principal")))
      // Log just to see if we get the value of the custom azp attribute, but imagine we are doing something meaningful here
      .doOnSuccess {
        logger.info("Principal info : name={}, azp={}",
          it.name,
          it.tokenAttributes["azp"]
        )

      }.then(next.handle(request))
  }
}